### DISCUSSION TITLE
Dealing with stateful processes created within a machine

### DISCUSSION THREAD

**Discussion by lewisl9029** [statelyai/xstate#3112] [Q&A] [answered]
Hi folks, I'm working on a vscode extension using xstate that needs to watch for file changes using the [createFileSystemWatcher ](https://code.visualstudio.com/api/references/vscode-api#workspace.createFileSystemWatcher) API (it has a similar API as [chokidar](https://github.com/paulmillr/chokidar), which people might be more familiar with).

I've been struggling to figure out the best way to model this part of the state machine, because the watcher returned from `createFileSystemWatcher` is a stateful process that maintains its own internal state and emits its own events, which means 2 things:

1) I need to store this stateful watcher process somewhere to interact with it later (to register/remove event handlers, and eventually clean up). The most obvious place is in context, but this snippet in the context docs makes me wonder if there's a better place to keep stateful processes like this that can change independently from the machine state:

> Never mutate the machine's context externally. Everything happens for a reason, and every context change should happen explicitly due to an event.

2) I need to somehow use the watcher as an event source (i.e. the onDidChange callbacks) and hook it into the machine so I can react to file change events and change the state of the machine. I'm a bit lost as to how to do this since I haven't found a way to send an event to the machine from within an event callback that might look something like this:

```js
// One of the many states of the machine that initializes the watcher
watching: {
  entry: xstate.assign({
    watcher: (context, event) => {
      if (context.watcher) return context.watcher
  
      const watcher = vscode.workspace.createFileSystemWatcher('./some-dir/**/*')

      watcher.onDidChange((uri) => {
        // Is it possible to send an event to the machine in response to this change event?
        // Or spawn an actor? Or something else that would let me translate these events into some change in machine state?
      })
      
      return watcher
    }
  })
}
```

The trickiest part is this watcher needs to be created within the machine inside one specific state. If it lived externally then I could just send events to the machine from the outside using service.send.

Would really appreciate it if folks could share some thoughts on how they would model a machine like this or point me towards some similar examples. Thanks!

**Comment by lewisl9029** [comment, marked as answer]
Following up on 2) just discovered that invoking callbacks seems to be the way to go: https://dev.to/mpocock1/xstate-why-i-love-invoked-callbacks-2f6i

Still unsure about 1), but will store a reference of the watcher in context unless/until I find a better way.

**Reply by Andarist** [reply]
Wrapping such a child process in an actor and invoking it or spawning it seems the way to go for me.
